corehq/celery.py

```diff
@@ -23,3 +27,25 @@
     run_checks()
 except AppRegistryNotReady:
     pass
+
+
+@after_task_publish.connect
+def celery_add_time_sent(headers=None, body=None, **kwargs):
+    info = headers if 'task' in headers else body
+    task_id = info['id']
+    cache.set('task.{}.time_sent'.format(task_id), datetime.datetime.utcnow())
+
+
+@task_prerun.connect
+def celery_record_time_to_start(task_id=None, task=None, **kwargs):
+    from corehq.util.datadog.gauges import datadog_gauge, datadog_counter
+    time_sent = cache.get('task.{}.time_sent'.format(task_id))
+    tags = [
+        'celery_task_name:{}'.format(task.name),
+        'celery_queue:{}'.format(task.queue),
+    ]
+    if time_sent:
+        time_to_start = (datetime.datetime.utcnow() - time_sent).total_seconds()
+        datadog_gauge('commcare.celery.task.time_to_start', time_to_start, tags=tags)
```

Comment:
By the way I want to share my choice of gauge here. A gauge is meant, conceptually, for tracking a single value over time—a classic example is disk usage. It's normally a poor choice for reporting timings—for example, request timings can within the same second vary between milliseconds and minutes—essentially it's a poor choice for capturing a sort of concurrent distribution of values. However, while on the surface this celery time-to-start metric may seem most similar to request timings, it's actually quite different in an important way. Since we're working with first-in-first-out queues, the time-to-start value for tasks that start in the same second is going to be very related, not bouncing around wildly between values, and there's no concurrency issue. A gauge metric only reports the last value within a flush period, which I think must be in the low seconds max. And for a use-case where the metric is only ever slowly rising or slowly falling, but generally continuous, that seems like a perfectly fine way to sample the data.

Reply:
👍 seems reasonable. I'm curious to see what the output looks like.

Reply:
So far it looks like this: https://app.datadoghq.com/dashboard/bdu-k5b-bz5/celery-overview?tile_size=m&page=0&is_auto=false&from_ts=1554111600000&to_ts=1554126000000&live=true&tpl_var_celery_queue=*&tpl_var_celery_task=*&fullscreen_widget=83087048

but that's hiding times larger than 5 minutes because of the cache expiry issue I fixed in https://github.com/dimagi/commcare-hq/pull/23833.